crates/s3_vectors/src/table.rs

```diff
@@ -0,0 +1,268 @@
+        write!(
+            f,
+            "{NAME} bucket={} index={}",
+            self.index.vector_bucket_name(),
+            self.index.index_name
+        )
+    }
+}
+
+#[async_trait]
+impl TableProvider for S3VectorsTable {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        Arc::clone(&self.schema)
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    async fn scan(
+        &self,
+        _state: &dyn Session,
+        _projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        let mut source = None;
+        for filter in filters {
+            if let Some(query_vector) = extract_query_vector(filter) {
+                let top_k = limit.unwrap_or(DEFAULT_TOP_K) as i32;
+                source = Some(Arc::new(QueryVectorsSource::new(
+                    Arc::clone(&self.client),
+                    self.index.clone(),
+                    self.schema(),
+                    query_vector,
+                    top_k,
+                )) as Arc<dyn DataSource>);
+                break;
+            }
+        }
+
+        let source = source.unwrap_or(Arc::new(ListVectorsSource::new(
+            Arc::clone(&self.client),
+            self.index.clone(),
+            self.schema(),
+        )));
+
+        Ok(Arc::new(DataSourceExec::new(source)))
+    }
+
+    fn supports_filters_pushdown(
+        &self,
+        filters: &[&Expr],
+    ) -> Result<Vec<TableProviderFilterPushDown>> {
+        Ok(filters
+            .iter()
+            .map(|f| {
+                if extract_query_vector(f).is_some() {
+                    TableProviderFilterPushDown::Exact
+                } else {
+                    TableProviderFilterPushDown::Unsupported
+                }
+            })
+            .collect())
+    }
+
+    async fn insert_into(
+        &self,
+        _state: &dyn Session,
+        input: Arc<dyn ExecutionPlan>,
+        _insert_op: InsertOp,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        let client = Arc::clone(&self.client);
+        let index = self.index.clone();
+        let schema = Arc::clone(&self.schema);
+
+        let sink = Arc::new(PutVectorsSink::new(client, index, schema));
+        let sort_order = None;
+        let exec = DataSinkExec::new(input, sink, sort_order);
+        Ok(Arc::new(exec))
+    }
+}
+
+fn contains_cosine_similarity_udf(expr: &Expr) -> Option<VectorData> {
+    if let Expr::ScalarFunction(ScalarFunction { func, args }) = expr {
+        if func.name() == "cosine_similarity" && args.len() == 2 {
+            if let Expr::Literal(ScalarValue::List(array), _) = &args[1] {
+                if let Some(list_array) = array.as_any().downcast_ref::<ListArray>() {
+                    let values = list_array.values();
+                    if let Some(float32_array) = values.as_any().downcast_ref::<Float32Array>() {
```

Comment:
pro-tip: Use `cast_to` with your desired target type to avoid coercions between e.g. f32 f64 and all the variants there